apps/emqx_cluster_link/src/emqx_cluster_link_api.erl

```diff
@@ -7,86 +7,454 @@
+    ).
+
+'/cluster/links/link/:name'(get, #{bindings := #{name := Name}}) ->
+    with_link(Name, fun(Link) -> handle_lookup(Name, Link) end, not_found());
+'/cluster/links/link/:name'(put, #{bindings := #{name := Name}, body := Params0}) ->
+    with_link(Name, fun() -> handle_update(Name, Params0) end, not_found());
+'/cluster/links/link/:name'(delete, #{bindings := #{name := Name}}) ->
+    with_link(
+        Name,
+        fun() ->
+            case emqx_cluster_link_config:delete(Name) of
+                ok ->
+                    ?NO_CONTENT;
+                {error, Reason} ->
+                    Message = list_to_binary(io_lib:format("Delete link failed ~p", [Reason])),
+                    ?BAD_REQUEST(Message)
+            end
+        end,
+        not_found()
+    ).
+
+'/cluster/links/link/:name/metrics'(get, #{bindings := #{name := Name}}) ->
+    with_link(Name, fun() -> handle_metrics(Name) end, not_found()).
 
 %%--------------------------------------------------------------------
 %% Internal funcs
 %%--------------------------------------------------------------------
 
+handle_list() ->
+    Links = get_raw(),
+    NodeResults = get_all_link_status_cluster(),
+    NameToStatus = collect_all_status(NodeResults),
+    EmptyStatus = #{status => inconsistent, node_status => []},
+    Response =
+        lists:map(
+            fun(#{<<"name">> := Name} = Link) ->
+                Status = maps:get(Name, NameToStatus, EmptyStatus),
+                maps:merge(Link, Status)
+            end,
+            Links
+        ),
+    ?OK(Response).
+
+handle_create(Name, Params) ->
+    case emqx_cluster_link_config:create(Params) of
+        {ok, Link} ->
+            ?CREATED(add_status(Name, Link));
+        {error, Reason} ->
+            Message = list_to_binary(io_lib:format("Create link failed ~p", [Reason])),
+            ?BAD_REQUEST(Message)
+    end.
+
+handle_lookup(Name, Link) ->
+    ?OK(add_status(Name, Link)).
+
+handle_metrics(Name) ->
+    case emqx_cluster_link:get_metrics(Name) of
+        {error, BadResults} ->
+            ?SLOG(warning, #{
+                msg => "cluster_link_api_metrics_bad_erpc_results",
+                results => BadResults
+            }),
+            ?OK(#{metrics => #{}, node_metrics => []});
+        {ok, NodeResults} ->
+            NodeMetrics =
+                lists:map(
+                    fun({Node, Metrics}) ->
+                        format_metrics(Node, Metrics)
+                    end,
+                    NodeResults
+                ),
+            AggregatedMetrics = aggregate_metrics(NodeMetrics),
+            Response = #{metrics => AggregatedMetrics, node_metrics => NodeMetrics},
+            ?OK(Response)
+    end.
+
+aggregate_metrics(NodeMetrics) ->
+    ErrorLogger = fun(_) -> ok end,
+    lists:foldl(
+        fun(#{metrics := Metrics}, Acc) ->
+            emqx_utils_maps:best_effort_recursive_sum(Metrics, Acc, ErrorLogger)
```

Comment:
I'd say we'll need to explain what those metrics mean.

Currently, the `routes` counter is tracked in _extrouter_, and thus means "number of routes pushed to me by `ClusterName`, aggregated by actor". Because different actors may push identical routes, this number will most of the time be larger than "number of routes `ClusterName` is interested in". I guess the last one is harder to measure because of the initial design goal of having leaderless route replication. To be clear it's still possible to measure through a (potentially expensive) full-scan of the extrouter table.

Reply:
> Because different actors may push identical routes, this number will most of the time be larger than "number of routes ClusterName is interested in

I thought duplications were handled in this code: https://github.com/emqx/emqx/blob/311419f6212b2a69233514d6366b70e273c8f09e/apps/emqx_cluster_link/src/emqx_cluster_link_extrouter.erl#L281-L299

Could there still be duplicates even with the check above, then?

Reply:
Oh wait. I misread that change initially, I think it _might be_ incorrect actually. Hope to be wrong, please help to verify.

```
    case MCounter band Marker of
        0 when OpName =:= add ->
            Res = mria:dirty_update_counter(?EXTROUTE_TAB, Entry, Marker),
            _ = emqx_cluster_link:routes_inc(ClusterName, 1),
```

^ This will bump counter each time an actor pushes a route not yet present in the extrouter table.
Two different actor will have different lanes, which means different `Marker` values, so them pushing the same route would cause 2 counter bumps.

```
            case mria:dirty_update_counter(?EXTROUTE_TAB, Entry, -Marker) of
                0 ->
                    Record = #extroute{entry = Entry, mcounter = 0},
                    ok = mria:dirty_delete_object(?EXTROUTE_TAB, Record),
                    _ = emqx_cluster_link:routes_inc(ClusterName, -1),
```

^ This deletion is actually could happen only when there are no more occupied lanes, which means that every actor had this route deleted. Moreover, this deletion might not happen at all, if there was concurrent `dirty_update_counter(+Marker)` in the meantime.

So those counter bumps are not actually symmetrical.

Reply:
I see, thanks for the explanation.  

Though, if there's only one actor per cluster link, it should still be safe to bump the counters, right?  Since the counter is scoped per cluster link name and `Marker`s are unique per actor?

Reply:
Number of actors per link is generally more than one, there are usually 1 or 2 actors per node. The idea is that each node replicates only its own route updates.

Reply:
Oh, true, there could be 2 actors (one for DS, the other for non-DS routes).  :see_no_evil: 

It sounds like the safest option then is to have a periodic full table scan that updates those metrics as gauges, then.

Reply:
Yeah. 🫠 I remember we had discussion with SergeTupchiy about the feasibility of having an extrouter table per cluster link, that would have probably helped.

I guess if we can tolerate small errors under high concurrency (in the "10 actors push the same route at the exact same moment to different target cluster nodes" sense), we can still rely on the result of `dirty_update_counter/3` and bump the counter if this was the first update, i.e. mirroring the current counter decrease logic in https://github.com/emqx/emqx/pull/13487/files/d9832252d8f1cafe30796fe555a4da088421359a#diff-2fe6367be77c076ca17ca1254e75b4a65a0de905a1e6af8badd52a347fe17455R296. I suspect that under really really high concurrency we risk bumping the counter several times on `add` operations and not decreasing it at all on `del` operations, but this assumption probably needs a double-check.

Reply:
One table per link would make it much simpler, indeed.  :smiley_cat: 

In my last commit, I added a bookkeeper process similar to what we have for persistent sessions.  WDYT?